docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,409 @@
+
+### IPAddressClaim
+
+An IPAddressClaim is a custom resource that can be used by infrastructure providers to allocate IP addresses from IP Pools that are managed by IPAM providers.
+
+### IPAddress
+
+An IPAddress is a custom resource that gets created by an IPAM provider to fulfil an IPAddressClaim, which then gets consumed by the infrastructure provider that created the claim.
+
+### IP Pool
+
+An IP Pool is a custom resource that is provided by a IPAM provider, which holds configuration for a single pool of IP addresses that can be allocated from using an IPAddressClaim.
+
+## Summary
+
+This proposal adds an API contract for pluggable IP Address Management to CAPI that any infrastructure provider can choose to implement. It allows them to dynamically allocate and release IP Addresses from IP Address Pools following a similar concept as Kubernetes' PersistentVolumes. Pools are managed by IPAM Providers, which allow to integrate with external systems like Netbox or Infoblox.
+
+## Motivation
+
+IP address management for machines is currently left to the infrastructure providers for CAPI. Most on-premise providers (e.g. CAPV) allow the use of either DHCP or statically configured IPs. Since Machines are created from templates, static allocation would require a single template for each machine, which prevents dynamic scaling of nodes without custom controllers that create new templates.
+
+While DHCP is a viable solution for dynamic IP assignment which also enables scaling, it can cause problems in conjunction with CAPI. Especially in smaller networks rolling cluster upgrades can exhaust the network in use. When multiple machines are replaced in quick succession, each of them will get a new DHCP lease. Unless the lease time is very short, at least twice as many IPs as the maximum number of nodes has to be allocated to a cluster.
+
+Metal3 has an ip-address-manager component that allows for in-cluster management of IP addresses through custom resources, allowing to avoid DHCP. CAPV allows to ommit the address from the template while having DHCP disabled, and will wait until it is set manually, allowing to implement custom controllers to take care of IP assignments. At DTAG we've extended metal3's ip-address-manager and wrote a custom controller for CAPV to integrate both with Infoblox, our IPAM solution of choice. The CAPV project has shown interest, and there has been a proposal to allow integrating metal3's ip-address-manager with external systems.
+
+All on-premise providers have a need for IP Address Management, since they can't leverage SDN features that are available in clouds such as AWS, Azure or GCP. We therefore propose to add an API contract to CAPI, which allows infrastructure providers to integrate with different IPAM systems. A similar approach to Kubernetes' PersistentVolumes should be used where infrastructure providers create Claims that reference a specific IP Pool, and IPAM providers fulfill claims to Pools managed by them.
```

Comment:
AWS also has infoblox extension, and announced their own VPC IPAM. If CAPA integrate with those in the future this contract could also be taken advantage
https://aws.amazon.com/marketplace/pp/prodview-opxe3p2cgudwe

Reply:
Wouldn't that be an integration with VPCs directly without the need to manually interact with Infoblox when creating a ec2 machine? Because in that case it wouldn't require this contract.

Reply:
Hm yeah you are right. That would be a detour. You can close this one.